DOCS-11147-cleanUpOperatorsDoc-duke (#1950)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/dw-operators.adoc

DataWeave supports the `update` operator, which enables you to update specified fields of a data structure with new values.

_Introduced in DataWeave 2.3.0. Supported by Mule 4.3 and later._

> Versions prior to Mule 4.3 require the following syntax to increase the `age` value in `myInput` by one without recreating the entire object:

Reviewer: possible rephrasing 

```suggestion
To increase the `age` value in `myInput` by one without recreating the entire object, versions prior to Mule 4.3 require the following syntax:
```

Author: Yeah, we jsut want to emphasize the version here by putting it in front because we are contrasting that with the previous entry that involves what you had to do prior to v4.3.

[source,weave,linenums]
----
%dw 2.0
var myInput = {
    "name": "Ken",
    "lastName":"Shokida",
    "age": 30
}

> output application/json

Reviewer 2: Just to follow other examples structures in the doc, maybe the output directive should be above the var directive.

Author: In the function docs, we usually put the output directive at the end, so I'll make these consistent with that practice on this page.

---
myInput mapObject ((value,key) ->
    if(key as String == "age")
       {(key): value as Number + 1}
    else
       {(key): value} )
----

> In Mule 4.3 and later, the `update` operator simplifies the syntax:

Author: ```suggestion
In Mule 4.3, the `update` operator simplifies the syntax:
```

[source,weave,linenums]
----
%dw 2.0

> var myInput = {
>     "name": "Ken",
>     "lastName":"Shokida",
>     "age": 30
> }
> output application/json

Reviewer 2: ```suggestion
output application/json
var myInput = {
    "name": "Ken",
    "lastName":"Shokida",
    "age": 30
}
```

}
----

* `value_to_update` represents the original value to update.

> * `update_expression` is the selector expression that matches a value in the `value_to_update`. 

Author: ```suggestion
* `update_expression` is the selector expression that matches a value in `value_to_update`. 
```

* `variable_name` is the name of the variable to bind to the matched `update_expression` value.
* `new_value` is the expression with the new value.

> * `[if(<conditional_expression>)]?` If the `conditional_expression` returns `true`, the script updates the value. This condition is required only for an optional update.

Author: ```suggestion
* `[if(<conditional_expression>)]?` If `conditional_expression` returns `true`, the script updates the value. This condition is required only for an optional update.
```

> * `[!]` marks the selector as an upsert. If the expression doesn't match anything, marking an expression with `!` makes the `update` operator create all the required elements and insert the new value.

Author: ```suggestion
* `[!]` marks the selector as an upsert. If the expression doesn't find a match, the `!` makes the `update` operator create all the required elements and insert the new value.
```

        "zipCode": "AB1234"
    }
}

> output application/json

Reviewer 2: Same suggestion here. Output  directive above var.

==== Index Selector

> This matching selector allows to match any array by its index.

Reviewer: this wording doesn't seem correct, can you verify. allows what to match?

Author: fixed here and in another sentence like it

> This matching selector allows to match any array by its index.

Author: ```suggestion
This matching selector enables you to match any array by its index.
```

==== Attribute Selector

> This matching selector allows to match any attribute value.

Author: ```suggestion
This matching selector enables you to match any attribute value.
```

==== Dynamic Selector

> Any selector can incorporate an expression that makes the selection dynamic. 

Author: ```suggestion
Any selector can incorporate an expression to make the selection dynamic. 
```

> The following example selects and dynamically updates a value of a DataWeave variable. The field is not hardcoded. The example uses `update` to change the value `name` to `Shoki`:

Reviewer 2: ```suggestion
The following example dynamically selects and updates a value where the field to be updated is not hardcoded, it could change at runtime. The example uses `update` to change the field with key `name`, obtained from the variable `theFiledName` and it changes its value to `Shoki`:
```

Author: done, copyedited

> The following example selects and dynamically updates a value of a DataWeave variable. The field is not hardcoded. The example uses `update` to change the value of the variable `theFieldName`, `name`, with a new value, `Shoki`:

Author: ```suggestion
The following example selects and dynamically updates a value of a DataWeave variable. The field is not hardcoded. The example uses `update` to change the value `name` to `Shoki`:
```

> The following example dynamically selects and updates the value of a field that is not hardcoded. Such a field can change at runtime. The example uses the variable `theFiledName` to obtain the field with key `name`, and it changes the value of that field to `Shoki`.

Author: ```suggestion
The following example dynamically selects and updates the value of a field that is not hardcoded. Such a field can change at runtime. The example uses the variable `theFieldName` to obtain the field with key `name`, and it changes the value of that field to `Shoki`.
```

=== Upserting

The `update` operator enables you to insert a field that is not present already by using the `!` symbol at the end of the selector expression. When the field is not present, the operator binds a `null` value to the variable. 

> The following example appends the specified key-value pair to the first object in the `myInput` array:

Reviewer 2: ```suggestion
The following example iterates through the objects in the Array and for each one updates the field `name` based on if it present or not. If it is not present, it will add the field `name` with the value "JHON" and if it is present, it will upper the name value.
```

Author: done, copyedited

.Source
[source,weave,linenums]
----
%dw 2.0

> var myInput = [{"lastName": "Doe"}, {"lastName": "Parker", "name": "Peter" }]

Reviewer 2: Suggestion: Output directive above variable.

=== Sugar Syntax

> The `update` operator enables you to use a default variable `$` that avoids the need to use a longer, standard expression.

Author: ```suggestion
The `update` operator enables you to use a default variable `$` that avoids the need to use the "boilerplate" equivalent, which is longer.
```

=== Sugar Syntax

> The `update` operator enables you to use the default variable `$` instead of the "boilerplate" equivalent, which is longer.

Reviewer 2: ```suggestion
The update operator allows you to write a shorter version syntax avoiding the `<variable_name> at` syntax by using a default variable named `$`.
```

Author: done, copyedited

> The `update` operator enables you to use a default variable `$` that avoids the need to use the "boilerplate" equivalent, which is longer.

Author: ```suggestion
The `update` operator enables you to use the default variable `$` instead of the "boilerplate" equivalent, which is longer.
```

> The following example uses `$` to select the `age` value instead of the longer equivalent shown in comments:

Reviewer 2: ```suggestion
The following example binds the value of the expression `.age` to the default variable name `$` used to express the new value, instead of the longer equivalent shown in comments:
```

Author: done, copyedited
